## CS431/631 Data Intensive Distributed Analytics
### Fall 2020 - Assignment 0
---

**Please edit this (text) cell to provide your name and UW student ID number!**
* **Name:** Jiujiu Ma
* **ID:** 20761836

---
For this assignment, you will be using Python to do some analysis of the text of Shakespeare's plays.    You should already have uploaded the text file (`Shakespeare.txt`) to the hub.   You should also have downloaded the Python tokenizer module, as `simple_tokenize.py`.

Let's first try running some simple Python code to make sure that everything is set up properly and ready to go.   The code in the next box will open `Shakespeare.txt`, read the file line by line, and tokenize each line that it reads in.    Try running the code by selecting the box and typing `Shift-Return`, i.e., hit the carriage return key while you are holding the shift key.   It may take a few seconds to run.

In [4]:
# this imports the SimpleTokenize function from the simple_tokenize.py file that you uploaded
from simple_tokenize import simple_tokenize

# Now, let's tokenize Shakespeare's plays
filename = 'Shakespeare.txt'""
with open(filename) as f:
    for line in f:
        # tokenize, one line at a time
        t = simple_tokenize(line)
t

['the', 'end']

---
### Important

The questions that follow ask you to implement functions whose prototypes are given to you. **Do _not_ change the prototypes of the functions. Do _not_ write code outside of the functions.** In particular, do not use the same cell as the function declaration invoke the function.

You may use specific cells, identified by `# Your tests here`, for test purposes. Code in these cells will *not* be executed when marking your assignment.

---

#### Question 1  (2/10 marks):
When the code is being executed, the cell number will change to \*, i.e., you should see `In [*]` in the left margin next to the cell.   After the code has finished running, the cell number will change to `In [1]` (indicating that this is the first code cell to be executed) and the notebook will display the resulting output immediately after cell.   In this case, you shoueld see the output
`['the', 'end']`.   In the next cell, briefly explain why the code produced this output.


#### Your answer to Question 1:
The varaible t gets overwritten for every line, and re.findall returns a list of the lower case of strings consisting of letters only. The last line of the text is "THE END", thus ['the', 'end'] is produced.





Remember that when you close and halt a notebook, any unsaved work in the notebook will be lost.   To save the contents of your notebook, use `Save and Checkpoint` (from the File menu).   

---

Now it is time for you to write some code.   Let's find the most frequently appearing tokens in Shakespeare's work.

#### Question 2 (4/10 marks):
In the next box, implement the function `top_50_tokens` using Python code that return the list of the 50 most frequently appearing tokens and their frequency, i.e., the number of times that each occurs.   Please use the `simple_tokenize` function, without modification, to tokenize the text, so that everyone is working with the same definition of what a token is.   If you wish, feel free to start with the Python code in the box above - just copy it from there and paste it below.

In [16]:
from simple_tokenize import simple_tokenize
import copy


def top_50_tokens(filename='Shakespeare.txt'):
    top_50_tokens_list = []
    word_list = []
    cnt_list = [] #temp lists for storing words and counts
    
    with open(filename) as f:
        
        for line in f:
        # tokenize, one line at a time
            t = simple_tokenize(line)
            for token in t:
                if token not in word_list:   # store new word and the count 1
                    word_list.append(token)
                    cnt_list.append(1)
                else:
                    index = word_list.index(token) # word exists, increment count
                    cnt_list[index] += 1

        top_50_word = word_list[:50].copy()
        top_50_cnt = cnt_list[:50].copy()
        count = len(cnt_list)
        assert(len(word_list) == count)
        assert(count > 50)
        for i in range(50, count - 1): # i = the index of the word of larger count
            for j in range(len(top_50_cnt)): # j = the index of existing word that needs to be replaced
                if cnt_list[i] > top_50_cnt[j]:
                    top_50_word[j] = word_list[i]
                    top_50_cnt[j] = cnt_list[i]
                    break

        for k in range(len(top_50_cnt)):
            top_50_tokens_list.append([top_50_word[k], top_50_cnt[k]]) #pair the word and its count
  
    return top_50_tokens_list

top_50_tokens()

[['the', 27378],
 ['and', 26082],
 ['i', 20717],
 ['you', 13630],
 ['is', 9137],
 ['not', 8512],
 ['me', 7777],
 ['your', 6657],
 ['have', 5885],
 ['no', 3797],
 ['do', 3753],
 ['that', 10915],
 ['lord', 3062],
 ['our', 3061],
 ['king', 2871],
 ['sir', 2763],
 ['them', 1980],
 ['man', 1876],
 ['but', 6277],
 ['as', 5744],
 ["i'll", 1745],
 ['us', 1620],
 ['tis', 1408],
 ['these', 1323],
 ['his', 6859],
 ['take', 1198],
 ['duke', 1071],
 ['exeunt', 1035],
 ['lady', 866],
 ['thou', 5491],
 ['ay', 768],
 ['to', 19661],
 ['prince', 743],
 ['own', 771],
 ['into', 639],
 ['eyes', 685],
 ['comes', 616],
 ['thy', 4034],
 ['noble', 615],
 ['gloucester', 573],
 ['with', 7778],
 ['henry', 564],
 ['caesar', 527],
 ['antony', 509],
 ['richard', 501],
 ['a', 14723],
 ['york', 486],
 ['where', 1235],
 ['falstaff', 444],
 ['warwick', 411]]

Be sure to test the code that you have written by running it.   When you submit your notebook to us, we will run your code when we mark you assignment.   As a sanity test on you output, our reference implementation finds that the most frequent word is "the", which occurs 27378 times.

---

Once you have found the 50 most frequent tokens, let's move on to something slightly more complicated.

#### Question 3 (4/10 marks):

Instead of the most frequent tokens appearing in Shakespeare's works, suppose that we want a list of words that appear after the word "perfect", on the same line, in Shakespeare's text. 
(Note: the "words" we are interested in for this question are tokens, as returned by simple_tokenize.)

For example, *All's Well That Ends Well* includes the line
>  Ere I can perfect mine intents, to kneel.

so "mine" should be part of the output, since it follows "perfect" on this line.  To keep the output from getting too long, include only words that appear after "perfect" on more than one line.

In the next box, implement the function `perfect_x` that returns a dictionary of key/value pairs, where the keys are the words that follow perfect on more than one line, and the values the number of lines in which the pattern is observed. For instance, if 'x' follows 'perfect' on 3 different lines, the entry in the dictionary will be ('x':3) As a sanity check on your output, our reference implementation finds 5 such words.

In [17]:
from simple_tokenize import simple_tokenize

def perfect_x(filename = 'Shakespeare.txt'):
    perfect_x_list = []
    word_list = []
    cnt_list = []

    with open(filename) as f:
        for line in f:
            # tokenize, one line at a time
            t = simple_tokenize(line)
            if "perfect" in t:
                for i in range(len(t)):
                    if t[i] == "perfect" and i != len(t) - 1: #emit if "perfect" is the last word in a line
                        perfect_x = t[i+1]
                        if perfect_x not in word_list:   # store new word and the count 1
                            word_list.append(perfect_x)
                            cnt_list.append(1)
                        else:  # word exists, increment count
                            index = word_list.index(perfect_x)
                            cnt_list[index] += 1

    for j in range(len(cnt_list)):
        if cnt_list[j] > 1:
            perfect_x_list.append([word_list[j], cnt_list[j]]) #pair the word and its count

    return perfect_x_list

perfect_x()


[['honour', 2], ['in', 4], ['love', 4], ['yellow', 2], ['that', 2]]

In [ ]:
# Your tests here

That's it!   Don't forget to save your work before closing and halting your notebook.      

When you are finished and ready to submit your assignment, download your notebook file (.ipynb) from the hub to your machine, and then follow the submission instructions in the assignment.